docs/dev_plan.md

```diff
@@ -0,0 +1,323 @@
+        direction LR
+        solver1{{Type 1}}
+        solver2{{Type 2}}
+        solver3{{Type 3}}
+    end
+
+    Output[[Output Model]]
+
+    subgraph Export
+        direction LR
+        export1{{Visualization}}
+        export2{{ASCII}}
+        export3{{Commercial}}
+    end
+
+    Finish(( ))
+
+    Start --- I/O
+    I/O --- DataModel
+    DataModel --- Solver
+    Solver --- Output
+    Output --- Export
+    Export --- Finish
+```
+
+It is critical to the sustainability and stability of `OpenTurbine`
+to maintain independence from external software even though there
+will be reliance on existing libraries for common tasks. The modular
+design should include data structures and API's that are general enough
+to support integration of third party libraries as well as the ability
+to change any library for another that accomplishes a similar task
+even if by alternative methods.
+
+### Performance
+
+A key design consideration of `OpenTurbine` is computational efficiency
+or performance. Both the quantity of work and the efficiency of data should
+be considered and measured (i.e. profiled) during any development effort.
+A modular architecture should support offloading computationally expensive
+tasks to hardware accelerators or specialized libraries, and support
+multiple options for doing so depending on user configurations and
+the computational environment.
+
+Similar to modularity in the architecure, expensive tasks should be
+structured in a kernel form. This low-level design pattern combines
+expensive mathematical operations into an aggregate form, and structures
+them so that performance libraries or compilers can paralellize the
+computation. This pattern helps to encapsulate expensive operations and
+algorithms. Additionally, it follows the modular architecture design
+described above in that it supports swapping accelerators or parallelization
+methods../
+
+`OpenTurbine` developers designing new algorithms and data structures
+should become familiar with the idea of
+[data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design).
+The key concept of this paradigm is to structure data so that it maps
+closely to the form it will be represented and used within the relevant
+processing units. For example, it is a common mistake in object-oriented
+design to represent data as arrays of structures (AoS),
+and this can lead to inefficiency in CPU cache
+usage since aggregate operations such as matrix math must load
+data from non-contiguous memory rather than perform the operations
+in a vector form. In a data-oriented design, data is more
+commonly reprented as structures of arrays (SoA) where contiguous
+portions of memory are loaded into the cache and operated on
+either by the CPU or a hardware accelerator. The graphic below
+illustrates the difference between structures of arrays and arrays
+of structures.
+
+```{image} ./static/AoS_SoA.pdf
+:alt: aos_soa
+:width: 400px
+:align: center
+```
+
```

Comment:
The above figure is not showing up for me. We might need to add the figure to the directory.

![image](https://user-images.githubusercontent.com/59463700/231824765-33f5a67c-b607-4b18-83a5-3823d159341d.png)

Reply:
Thanks for fixing the figure. It looks like a great addition to the description here. Just a minor suggestion, would it add to the clarity of the figure if we added data types to the array elements such as done [here](https://stackoverflow.com/questions/8377667/is-the-memory-allocated-for-struct-members-continguous-what-if-a-struct-member/8377717#8377717)?

Reply:
Sure I'm happy to add these details. However, there's a nuance that's worth clarifying. I've understood the benefit from SoA to be that sections of an array can be loaded to the cache at a time, and the contiguity being that many adjacent elements of an array are used in a computation (thus leading to SIMD). In the example described in that SO post, they're suggesting that avoiding the padding between different data types is the benefit. I suppose it's a combination of both in the end.

Reply:
Ok I just read the SO post you linked to in the text (https://stackoverflow.com/questions/17924705/structure-of-arrays-vs-array-of-structures) and that clarifies it a bit more.

Reply:
Looks good! Thanks for taking the time to expand on the concept.